src/lib/sample_sheet.rs

```diff
@@ -265,105 +321,116 @@ impl SampleSheet {
         let mut ordinal = 1;
         let mut samples: Vec<SampleMetadata> = vec![];
-        while line_index < records.len() {
+        for record in &records[1..] {
             // allow an empty line
-            // TODO: allow a line with the correct # of columns but all empty lines
-            if records[line_index].is_empty() {
+            if record.is_empty() {
                 line_index += 1;
                 continue;
             }
             // make sure we have the correct number of columns
-            if data_header.len() != records[line_index].len() {
+            if header.len() != record.len() {
                 return Err(SampleSheetError::SampleInvalidNumberOfColumns {
-                    actual: records[line_index].len(),
-                    expected: data_header.len(),
+                    actual: record.len(),
+                    expected: header.len(),
                     line_number: line_index + 1,
-                    line: records[line_index].into_iter().join(","),
+                    line: record.into_iter().join(","),
                 });
             }
 
             // Build the sample and convert it to SampleMetadata
-            let sample: SampleMetadata =
-                records[line_index].deserialize(Some(data_header)).map_err(|e| {
-                    SampleSheetError::SampleInvalidLine { source: e, line: line_index + 1 }
-                })?;
-            let sample_metadata: SampleMetadata = sample.update_with(ordinal, line_index + 1)?;
+            let sample: SampleMetadata = record.deserialize(Some(header)).map_err(|e| {
+                SampleSheetError::SampleInvalidLine { source: e, line: line_index + 1 }
+            })?;
+            let sample_metadata: SampleMetadata =
+                sample.update_with_and_set_demux_barcode(ordinal, line_index + 1)?;
             samples.push(sample_metadata);
 
             ordinal += 1;
             line_index += 1;
         }
+        if samples.is_empty() {
+            return Err(SampleSheetError::NoSamples);
+        }
+
+        Ok(samples)
+    }
+
+    /// Reads from a Sample Sheet represented as a vector if `StringRecord`s.  Skips over lines
+    /// optionally until the `[Demux]` section is found, upating the given command line options,
+    /// and then the `[Data]` section, to parse a list of samples.  Validates the samples are
+    /// unique and their barcodes are dissimilar enough.
+    fn from_sample_sheet_string_records(
+        records: &[StringRecord],
+        opts: Opts,
+    ) -> Result<Self, SampleSheetError> {
+        if records.is_empty() {
+            return Err(SampleSheetError::Empty);
+        }
+
+        // Parse demultiplexing settings ([Demux] section)
+        let opts: Opts = {
+            match SampleSheet::find_section(records, "[Demux]") {
+                None => Ok(opts),
+                Some((start, end)) => {
+                    SampleSheet::parse_and_update_demux_options(&records[start..=end], opts)
```

Comment:
docs say find_section returns zero based exclusive.  Your use of `..=` here says that is a lie.

Reply:
It is sorry